### PR TITLE
Consolidate docs substitutions

### LINKED ISSUE #11063: RST Substitutions in docstrings
## Description

In ``docs/conf.py`` it is possible to add often used substitutions to the ``rst_epilog`` variable. These substitutions can **also be used in docstrings**.  I propose we replace many of the docstrings with their RST substitution counterparts, at least for very obvious substitutions like ``Unit`` and ``Quantity`` and ``SkyCoord``, etc.

As @mhvk  points out in https://github.com/astropy/astropy/pull/10662#discussion_r527127005, we can define things like "Quantity-like" where a Column with a unit will work where a Quantity would too. Some of these, like where a string can replace a unit, might be a multiple RST replacement (see example below).

- - -

## Example

docs/conf.py
```
rst_epilog += """

.. |Unit| replace:: :class:`~astropy.units.Unit`
.. |Unit.to| replace:: :meth:`~astropy.units.Unit.to`  # descriptive
.. |Quantity| replace:: :class:`~astropy.units.Quantity`
.. |SkyCoord| replace:: :class:`~astropy.coordinates.SkyCoord`

.. |find_api_page| replace:: :func:`~astropy.find_api_page`  # function

.. |Representation| replace:: :class:`~astropy.coordinates.BaseRepresentation`  # simplification
.. |Differential| replace:: :class:`~astropy.coordinates.BaseDifferential`  # simplification

.. |Quantity-like| replace:: Quantity-like  # replacement invariant
.. |unit-like| replace:: :class:`~astropy.units.Unit` or str  # multiple subs

"""
```

some_file.py
```python
def func(x):
    """function

    Parameters
    ——————
    x : |Quantity|
    """
    pass

```

### PR DESCRIPTION
**Pull request by nstarman**
### Description

The docs repeatedly define ReST substitutions for Astropy classes like |Quantity| or |Table|. This PR consolidates them into the ``docs/conf.py::rst_epilog`` so substitutions need only be defined once, and can also be used in code docstrings.

Fixes #11063

### CODE REVIEW

**Comment by nstarman** [comment]
@pllim. Purely cosmetic, so push to next milestone?

**Comment by nstarman** [comment]
Alternatively, I can definitely do this for Friday if it can be pushed in .

**Comment by pllim** [comment]
This looks uncontroversial enough that it might still make it into 4.3. Though, we might want to document that this is how we want the refs to be done in the dev doc for... doc.

**Comment by nstarman** [comment]
I'll add a note to that regard.

**Comment by nstarman** [comment]
Darn. Pytest doc sub is flaky. Will try to fix.

File: docs/table/access_table.rst [hunk 1]
```diff
@@ -178,7 +176,7 @@ Accessing Properties
 
 The code below shows accessing the table columns as a |TableColumns| object,
 getting the column names, table metadata, and number of table rows. The table
-metadata is an ordered dictionary (OrderedDict_) by default.
+metadata is an ordered dictionary (|OrderedDict|) by default.
```
Comment (mhvk) [inline docs/table/access_table.rst:179]: This one I'm not sure off; why not just `:class:`~collections.OrderedDict`? It does not seem to be used anywhere else, and we're attempting to remove its use...

File: docs/table/construct_table.rst [hunk 2]
```diff
@@ -691,9 +689,9 @@ meta
 
 The ``meta`` argument is an object that contains metadata associated
 with the table. It is recommended that this object be a dict or
-OrderedDict_, but the only firm requirement is that it can be copied with
+|OrderedDict|, but the only firm requirement is that it can be copied with
```
Comment (mhvk) [inline docs/table/construct_table.rst:692]: So, here and just below would be the other places to replace with :class:`~collections.OrderedDict`

**Review by mhvk — COMMENTED** [review]
Another really nice clean-up action! I have only one request, which is not to include `OrderedDict` in the list - maybe the best argument is just to stick to astropy classes only...

**Comment by mhvk** [comment]
@taldcroft, @adrn - as two "randomly" chosen others to have a look at this: do you agree with the concept of centralizing aliases used in the docs? I've reviewed this and the implementation is all OK with me (modulo a small change easily fixed), so if you agree that the idea is sound, I'll merge.

**Comment by adrn** [comment]
I agree with the concept! Unfortunately don't have time to review this in detail today, maybe tomorrow...

**Comment by taldcroft** [comment]
I agree with the concept.

**Comment by mhvk** [comment]
@adrn, @taldcroft - no need to review, as I think I was thorough enough; the agreement with the concept was what I was looking for!

@nstarman - ping me when you removed the `OrderedDict` short-cut and we'll get this in! But also do check the `readthedocs` failure - looks related!

File: docs/table/access_table.rst (hunk 1 above)
Comment (nstarman) [reply docs/table/access_table.rst:179]: ✅

File: docs/table/construct_table.rst (hunk 2 above)
Comment (nstarman) [reply docs/table/construct_table.rst:692]: ✅

**Comment by nstarman** [comment]
@mhvk. Ok. I finally managed to resolve the RTD issues with pytest. Turns out it was very similar to https://thomas-cokelaer.info/blog/2020/04/sphinx-raise-a-warning-duplicate-explicit-target-name/. I'm going to be pushing my changes a few pieces at a time to make sure I resolved the behavior. I'll ping you when it's all done!

**Comment by nstarman** [comment]
Hunting down that problem was a fair bit of work because adding subs to ``rst_epilog`` appends them to every file and ``tox -e build_docs`` doesn't show you the specific file that raises the error when rst_epilog has the sub and the file has a non-anonymous link of the same name. I'm trying to find a place in the documentation to add a note about this. @mhvk, any suggestions?

**Comment by nstarman** [comment]
Ok, @mhvk, except for a possible note in the dev docs, this is G2G!

**Comment by nstarman** [comment]
I see a lot of potential for followup:

- More link consolidation. I was pretty conservative: only collapsing ReST substitutions and transcluded files.
- Replacing near-duplicated links with their ``rst_epilog`` counterparts.
- using the links in docstrings. I almost successfully avoided touching anything not in the docs.

**Comment by mhvk** [comment]
@nstarman - OK, let's get this in! Probably good to have separate issue(s) for the follow-up. My overall sense is to keep it relatively limited, as you have it here, and otherwise err on the side of having references close to where they are needed.

**Review by mhvk — APPROVED** [review]
Great, thanks!!!

**Comment by embray** [comment]
Thank you for doing this.  I wonder if we could move all of these into a separate file from `conf.py` though?  I feel like it's a big chunk of the file now and makes it harder to read.  It's also not super clear that it's a place now to put common links to use throughout the docs.

I propose moving it out to something like `docs/common_links.txt` that can be read in by conf.py.

**Comment by mhvk** [comment]
Separating it out sounds good.

**Comment by nstarman** [comment]
@embray . I like the idea. I had kept the substitutions grouped by subpackages, but still all squashed together. This will allow them to be a little more separate.

**Comment by nstarman** [comment]
@embray @mhvk, I opened #11624

**Pull request merged**